python/pylibcudf/pylibcudf/contiguous_split.pyx

```diff
@@ -111,6 +127,165 @@ cdef class PackedColumns:
+        out.table = input
+        out.mr = temp_mr
+        out.stream = stream
+        out.c_obj = move(obj)
+        return out
+
+    cpdef bool has_next(self):
+        """
+        Check if the packer has more chunks to pack.
+
+        Returns
+        -------
+        True if the packer has chunks still to pack.
+        """
+        with nogil:
+            return dereference(self.c_obj).has_next()
+
+    cpdef size_t get_total_contiguous_size(self):
+        """
+        Get the total size of the packed data.
+
+        Returns
+        -------
+        Size of packed data.
+        """
+        with nogil:
+            return dereference(self.c_obj).get_total_contiguous_size()
+
+    cpdef size_t next(self, DeviceBuffer buf):
+        """
+        Pack the next chunk into the provided device buffer.
+
+        Parameters
+        ----------
+        buf
+            The device buffer to pack into, must be the same size as
+            the `user_buffer_size` used to construct the packer.
+
+        Returns
+        -------
+        Number of bytes packed.
+        """
+        cdef device_span[uint8_t] d_span = device_span[uint8_t](
+            <uint8_t *>buf.c_data(), buf.c_size()
+        )
+        with nogil:
+            return dereference(self.c_obj).next(d_span)
+
+    cpdef memoryview build_metadata(self):
+        """
+        Build the metadata for the packed representation.
+
+        Returns
+        -------
+        memoryview of metadata suitable for passing to `unpack_from_memoryviews`.
+        """
+        cdef unique_ptr[vector[uint8_t]] metadata
+        with nogil:
+            metadata = move(dereference(self.c_obj).build_metadata())
+        return memoryview(HostBuffer.from_unique_ptr(move(metadata)))
+
+    cpdef tuple pack_to_host(self, DeviceBuffer buf):
+        """
+        Pack the entire table into a host buffer.
+
+        Parameters
+        ----------
+        buf
+           The device buffer to use as a staging buffer, must be the
+           same size as the `user_buffer_size` used to construct the
+           packer.
+
+        Returns
+        -------
+        tuple of metadata and packed host data (as memoryviews)
+
+        Notes
+        -----
+        This is stream-ordered with respect to the stream used when
+        creating the `ChunkedPack` and syncs that stream before returning.
+        """
+        cdef size_t offset = 0
+        cdef size_t size
+        cdef device_span[uint8_t] d_span = device_span[uint8_t](
+            <uint8_t *>buf.c_data(), buf.c_size()
+        )
+        cdef cudaError_t err
+        cdef unique_ptr[vector[uint8_t]] h_buf = (
+            make_unique[vector[uint8_t]](
+                dereference(self.c_obj).get_total_contiguous_size()
+            )
+        )
+        cdef cudaStream_t stream = self.stream.view().value()
+        with nogil:
+            while dereference(self.c_obj).has_next():
```

Comment:
deference this once?

Reply:
This doesn't work because `chunked_pack` has an implicitly deleted copy-ctor.